### PR TITLE
Parameterised tests using GoogleTest

### LINKED ISSUE #3: Parameterised tests
•	Parametrized tests and data-driven testing (Introduction).
•	Examples
•	Combining test fixtures with parametrised tests

### PR DESCRIPTION
**Pull request by lkr-ind**
This PR addresses the Close #3. 

I have used the classes and concepts introduced in #14 for consistency.

### CODE REVIEW

File: _episodes/l4_04_parameterized_tests.md
Comment (dalonsoa) [inline _episodes/l4_04_parameterized_tests.md:1]: Filename should be `l1_04_parametrize_tests.md`. Changing the value of `l` puts the episode in a different session in the schedule that appears in the `index`. See the structure of files in https://github.com/ImperialCollegeLondon/intermediate_grad_school_git_course/tree/gh-pages/_episodes where the course is split in two sessions of 2h each, therefore files starting with `l1` and `l2`.

File: _episodes/l4_04_parameterized_tests.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+- "We learnt how to create our own parameterised tests."
+- "We learnt how to combine test fixture with parameterised test."
+---
+
+## 1. Introduction to Parameterised Tests
+**Parameterized tests**, also known as **data-driven tests**, are a feature provided by testing frameworks like Google Test that allows us to write a single test case that can be executed with different sets of test data or inputs. Instead of duplicating similar test cases with slight variations, parameterized tests enable us to define a test once and run it with multiple inputs or test data.
+
+In order to understand the importance of parameterised tests and why do we need them, let us consider a very small example. For this chapter, we will be using our `Employee` class that we created in last chapter.
+
+Let us suppose that we want to test that net bonus calculation works fine for different number of years of expereince. Remember that our `Employee` class adds an additional bonus of $1000£$ when an employee has worked for more than 10 years. As a first approach, we might be tempted to write multiple tests for the same function using test fixtures in the same way we have been doing so far.
```
Comment (dalonsoa) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
Let us suppose that we want to test that net bonus calculation works fine for different number of years of experience. Remember that our `Employee` class adds an additional bonus of $1000£$ when an employee has worked for more than 10 years. As a first approach, we might be tempted to write multiple tests for the same function using test fixtures in the same way we have been doing so far.
```

File: code/Chapter4/1_Not_parameterised.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include <iostream>
```
Comment (dc2917) [inline code/Chapter4/1_Not_parameterised.cpp]:
```suggestion
```

File: code/Chapter4/1_Not_parameterised.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#include <gtest/gtest.h>
+#include "employee.h"
+
+using namespace :: testing;
+using namespace :: std;
```
Comment (dc2917) [inline code/Chapter4/1_Not_parameterised.cpp]:
```suggestion
```

File: code/Chapter4/2_Test_using_for_loop.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#include <iostream>
```
Comment (dc2917) [inline code/Chapter4/2_Test_using_for_loop.cpp]:
```suggestion
```

File: code/Chapter4/2_Test_using_for_loop.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+#include <gtest/gtest.h>
+#include "employee.h"
+
+using namespace :: testing;
+using namespace :: std;
```
Comment (dc2917) [inline code/Chapter4/2_Test_using_for_loop.cpp]:
```suggestion
```

File: code/Chapter4/3_Parameterised_not_using_fixture.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#include <iostream>
```
Comment (dc2917) [inline code/Chapter4/3_Parameterised_not_using_fixture.cpp]:
```suggestion
```

File: code/Chapter4/3_Parameterised_not_using_fixture.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#include <gtest/gtest.h>
+#include "employee.h"
+
+using namespace :: testing;
+using namespace :: std;
```
Comment (dc2917) [inline code/Chapter4/3_Parameterised_not_using_fixture.cpp]:
```suggestion
```

File: code/Chapter4/3_Parameterised_not_using_fixture.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+struct TestValues{
+    int input;
+    int output;
+    
+    //construtor of values struct
+    TestValues(int in, int out) : input(in), output(out) {}
+};
+
+// Create a parameterised class by deriving from testing::TestWithParam<T> where T could be any valid C++ type.
+class EmployeeTestParameterised : public::TestWithParam<TestValues> {
```
Comment (dc2917) [inline code/Chapter4/3_Parameterised_not_using_fixture.cpp]:
```suggestion
class EmployeeTestParameterised : public::testing::TestWithParam<TestValues> {
```

File: code/Chapter4/4_param_test_based_fixture.cpp [hunk 9]
```diff
@@ -0,0 +1,75 @@
+#include <iostream>
```
Comment (dc2917) [inline code/Chapter4/4_param_test_based_fixture.cpp]:
```suggestion
```

File: code/Chapter4/4_param_test_based_fixture.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+#include <gtest/gtest.h>
+#include "employee.h"
+
+using namespace :: testing;
+using namespace :: std;
```
Comment (dc2917) [inline code/Chapter4/4_param_test_based_fixture.cpp]:
```suggestion
```

File: code/Chapter4/employee.cpp
Comment (dc2917) [inline code/Chapter4/employee.cpp:1]: Can employee.{cpp,h} be re-used from chapter 3?

File: _episodes/l4_04_parameterized_tests.md [hunk 11]
```diff
@@ -0,0 +1,353 @@
+---
+title: "Parameterised Tests using GoogleTest"
+teaching: TBC
+exercises: TBC
+questions:
+- "What is a parameterised test?"
+- "How to write your own parameterised tests using GoogleTest?"
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
- "How can I write my own parameterised tests using GoogleTest?"
```

File: _episodes/l4_04_parameterized_tests.md [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+objectives:
+- "Understand why do we need a parameterised test."
+- "Understand how to create a parameterised test using GoogleTest."
+- "Understand the advantages of parameterised tests."
+- "Understand how to combine a test fixtures with parameterised test."
+keypoints:
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md:13]: Again, the keypoints need more detail, rather than just re-stating the objectives

File: _episodes/l4_04_parameterized_tests.md [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+- "We learnt about parameterised tests in GoogleTest."
+- "We learnt how to create our own parameterised tests."
+- "We learnt how to combine test fixture with parameterised test."
+---
+
+## 1. Introduction to Parameterised Tests
+
+**Parameterized tests**, also known as **data-driven tests**, are a feature provided by testing frameworks like Google Test that allows us to write a single test case that can be executed with different sets of test data or inputs. Instead of duplicating similar test cases with slight variations, parameterized tests enable us to define a test once and run it with multiple inputs or test data.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
**Parameterised tests**, also known as **data-driven tests**, are a feature provided by testing frameworks like Google Test that allows us to write a single test case that can be executed with different sets of test data or inputs. Instead of duplicating similar test cases with slight variations, parameterised tests enable us to define a test once and run it with multiple inputs or test data.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+
+In order to understand the importance of parameterised tests and why do we need them, let us consider a very small example. For this chapter, we will be using our `Employee` class that we created in last chapter.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
In order to understand the importance of parameterised tests and why we need them, let us consider a very small example. For this chapter, we will be using our `Employee` class that we created in last chapter.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+
+Let us suppose that we want to test that net bonus calculation works fine for different number of years of experience. Remember that our `Employee` class adds an additional bonus of $1000£$ when an employee has worked for more than 10 years. As a first approach, we might be tempted to write multiple tests for the same function using test fixtures in the same way we have been doing so far.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
Let us suppose that we want to test that net bonus calculation works fine for different number of years of experience. Remember that our `Employee` class adds an additional bonus of £1000 when an employee has worked for more than 10 years. As a first approach, we might be tempted to write multiple tests for the same function using test fixtures in the same way we have been doing so far.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+    EXPECT_EQ(employee.getNetBonus(), 2000);
+}
+
+TEST_F(EmployeeTestFixture, NetBonusIsCorrectWhenYearsGreaterThan10) {
+    employee.setNumberYearsEmployed(15);
+    EXPECT_EQ(employee.getNetBonus(), 3000);
+}
+```
+
+While the above solution works pretty well, it has a serious drawback. If we carefully look at the tests, we see that the test logic is repeated in both the tests. The only difference between the two tests are the input and output values. Moreover, managing such test will become problematic as the number of test conditions (or input/output) values increases.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
While the above solution works pretty well, it has a serious drawback. If we carefully look at the tests, we see that the test logic is repeated in both the tests. The only difference between the two tests are the input and output values. Moreover, managing such tests will become problematic as the number of test conditions (or input/output) values increases.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+[----------] 1 test from EmployeeTestFixture (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 1 test from 1 test suite ran. (0 ms total)
+[  PASSED  ] 1 test.
+```
+
+Although, the `for` loop served our purpose and we were able to run our test for multiple values, there is a big problem in this approach. If we carefully look at the output, we can see that both (or multiple values if present) the test cases were combined into a single test. This violates the general rule that we should test only one thing in a test or one assertion per test.
+
+Moreover, the problem gets worse when one of the test fails. In order to understand what happens during a test failure when using a for loop, let us intentionally change the expected output value to an incorrect value. In file, [2_Test_using_for_loop.cpp](../code/Chapter4/2_Test_using_for_loop.cpp), you can make the following change.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
Moreover, the problem gets worse when one of the tests fails. In order to understand what happens during a test failure when using a for loop, let us intentionally change the expected output value to an incorrect value. In file, [2_Test_using_for_loop.cpp](../code/Chapter4/2_Test_using_for_loop.cpp), you can make the following change.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+[----------] Global test environment tear-down
+[==========] 1 test from 1 test suite ran. (0 ms total)
+[  PASSED  ] 0 tests.
+[  FAILED  ] 1 test, listed below:
+[  FAILED  ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears
+
+ 1 FAILED TEST
+```
+
+From the output, we can clearly see that it results in complete failure of the test even though one of the conditions (or test) was right. Moreover, the output does not help much to figure out which test has exactly failed.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
From the output, we can clearly see that it results in complete failure of the test even though one of the conditions (or tests) was right. Moreover, the output does not help much to figure out which test has failed.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+
+The solution for above mentioned issues is to make use of parameterised tests and the next section describes that.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
The solution to these issues is to make use of parameterised tests and the next section describes that.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+struct MyStruct{
+    int input;
+    int output;
+    
+    //construtor of values struct
+    MyStruct(int in, int out) : input(in), output(out) {}
+};
+```
+
+Once you have defined a structure to hold your values, you can create an instance of the same with the actual set of input and output values as shown below.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
Once you have defined a structure to hold your values, you can create an instance of it with the actual set of input and output values as shown below.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+
+```cpp
+MyStruct MyValues[] = {
+    MyStruct{InputVal1, OutputVal1},  //using constructor to create an instance of MyStruct.  
+    MyStruct{InputVal2, OutputVal2}
+};
+```
+
+3. **Create you test with TEST_P macro**: Instead of `TEST_F` macro that we used for test fixture, we use a `TEST_P` macro where `P` stands for `Parameterised` as shown below.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
3. **Create your test with TEST_P macro**: Instead of `TEST_F` macro that we used for test fixture, we use a `TEST_P` macro where `P` stands for `Parameterised` as shown below.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+
+```cpp
+TEST_P(YourTestParameterisedClass, NameofTest) {
+    // Test logic goes here.
+}
+```
+
+4. **Instantiate your test**: Finally, we instantiate our test by using `INSTANTIATE_TEST_SUITE_P` macro. The general syntax of this macro is given below.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
4. **Instantiate your test**: Finally, we instantiate our test by using the `INSTANTIATE_TEST_SUITE_P` macro. The general syntax of this macro is given below.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+
+```cpp
+INSTANTIATE_TEST_SUITE_P(SuitableNameTest, 
+                         YourTestParameterisedClass,
+                         ValuesIn(MyValues));
+```
+
+In above cell, the first argument to `INSTANTIATE_TEST_SUITE_P` could be any suitable name. GoogleTest will add this as a PREFIX to the test name when you will run the test. The second argument is the name of the parameterised class that you have created which is also the first argument for `TEST_P` macro. Finally, the last argument is a `ValuesIn()` function which is defined in GoogleTest library. It helps to inject the test values into the parameterised test one by one.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
In the above cell, the first argument to `INSTANTIATE_TEST_SUITE_P` could be any suitable name. GoogleTest will add this as a PREFIX to the test name when you will run the test. The second argument is the name of the parameterised class that you have created, which is also the first argument for `TEST_P` macro. Finally, the last argument is a `ValuesIn()` function which is defined in the GoogleTest library. It helps to inject the test values into the parameterised test one by one.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+[----------] 2 tests from NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 2 tests from 1 test suite ran. (0 ms total)
+[  PASSED  ] 2 tests.
+```
+
+In this output, there are two things worth noting:-
+
+1. As expected, we are now running two tests as compared to just one in case of a for loop.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
1. As expected, we are now running two tests as compared to just one in the case of a `for` loop.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+2. The test name `NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0` is a combination of the following:-
+    - A Prefix `NetBonusIsCorrectForDifferentYears` coming from INSTANTIATE_TEST_SUITE_P.
+    - Parameterised class name `EmployeeTestParameterised` coming from the first argument of `TEST_P` macro.
+    - Test name `NetBonusIsCorrectForDifferentYears` coming from the second argument of `TEST_P` macro.
+    - Finally, the iteration number.
+
+With this parameterised test, we were able to solve the issues that we were discussing above. However, in doing so, we changed the test fixture and converted it to use `TEST_P` macro. Our previous tests based on `TEST_F` macro will not work anymore as it has been replaced. The important question is: What shall we do so that we can still keep all our useful tests from test fixtures while still being able to add parameterised test. The solution is to combine test fixture with parameterised test and the next subsection explains that.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
With this parameterised test, we were able to solve the issues that we were discussing above. However, in doing so, we changed the test fixture and converted it to use `TEST_P` macro. Our previous tests based on `TEST_F` macro will not work anymore as it has been replaced. The important question is: What shall we do so that we can still keep all our useful tests from test fixtures while still being able to add parameterised test? The solution is to combine test fixtures with parameterised tests and the next subsection explains that.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+In order to create a parameterised test from a test fixture, all we need to do is to create a parameterised test class which derives from both the test fixture class and `testing::WithParamInterface<T>` class (defined in GoogleTest) to create parameterised tests.
+
+```cpp
+// create a parameterised test class from the fixture defined above.
+class YourParameterisedClass : public YourFixtureClass, 
+                               public WithParamInterface<T> {
+};
+```
+
+For the demonstration purpose, let us assume that we now want to check our tax calculation function `getTaxAmount()` which has more branches as compared to bonus calculation. For complete code, see the file [4_param_test_based_fixture.cpp](../code/Chapter4/4_param_test_based_fixture.cpp). We give a small section of code below for reference.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
For the purpose of demonstration, let us assume that we now want to check our tax calculation function `getTaxAmount()` which has more branches as compared to bonus calculation. For complete code, see the file [4_param_test_based_fixture.cpp](../code/Chapter4/4_param_test_based_fixture.cpp). We give a small section of code below for reference.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+    EXPECT_EQ(employee.getTaxAmount(), current_test_case_value.out_tax);
+}
+
+// Instantiate the test case with the values array.
+INSTANTIATE_TEST_SUITE_P( CheckTaxCalculation, 
+                          EmployeeTestParameterisedFixture,
+                          ValuesIn(values));
+```
+
+The major change as compared to our previous example is shown in the cell below and this change is responsible to generate a parameterised test using a test fixture.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
The major change as compared to our previous example is shown in the cell below and this change is responsible for generating a parameterised test using a test fixture.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+};
+```
+
+In addition, we used a function `GetParam()` defined in `gtest.h`. This function can help us to get the input values passed via `ValuesIn()` function and use it in the test logic according to our requirements. In this case, it helps us to retrieve 4 values in the order `inp_salary`, `inp_bonus`, `inp_years_employed` and `out_tax` for each test case. Thus, `GetParam()` provides a convenient way to retrieve multiple values and use them in our test logic.
+
+## 4. Advantages of Parameterised tests
+
+From the above discussion, we can see that the parameterised tests have the following advantages.
+
+1. **Code Reusability**: With parameterized tests, we can write a single test case that can be executed with different inputs or test data. This promotes code reusability by eliminating the need to duplicate similar test cases. Instead, we can define the test logic once and apply it to multiple scenarios, reducing code duplication and improving maintainability.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
1. **Code Reusability**: With parameterised tests, we can write a single test case that can be executed with different inputs or test data. This promotes code reusability by eliminating the need to duplicate similar test cases. Instead, we can define the test logic once and apply it to multiple scenarios, reducing code duplication and improving maintainability.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+
+2. **Increased Test Coverage**: Parameterized tests allows us to easily test a wide range of input values or test cases without writing separate test cases for each variation. This enables us to achieve better test coverage by covering various combinations, edge cases, and boundary values in a concise manner.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
2. **Increased Test Coverage**: Parameterised tests allow us to easily test a wide range of input values or test cases without writing separate test cases for each variation. This enables us to achieve better test coverage by covering various combinations, edge cases, and boundary values in a concise manner.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+
+3. **Simplified Test Maintenance**: When changes are required in the test logic, having parameterized tests simplifies the maintenance process. Instead of modifying multiple test cases individually, we only need to update the single test case, which will automatically be executed with the new test data. This saves time and effort in maintaining and updating the tests.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
3. **Simplified Test Maintenance**: When changes are required in the test logic, having parameterised tests simplifies the maintenance process. Instead of modifying multiple test cases individually, we only need to update the single test case, which will automatically be executed with the new test data. This saves time and effort in maintaining and updating the tests.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 31]
```diff
@@ -0,0 +1,353 @@
...
+
+4. **Simplified Test Reporting**: Parameterized tests provide a concise way to report test results for multiple test cases. Each instance of the parameterized test is reported individually, allowing us to identify which specific inputs or test data passed or failed. This facilitates quick identification and debugging of issues.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
4. **Simplified Test Reporting**: Parameterised tests provide a concise way to report test results for multiple test cases. Each instance of the parameterised test is reported individually, allowing us to identify which specific inputs or test data passed or failed. This facilitates quick identification and debugging of issues.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 32]
```diff
@@ -0,0 +1,353 @@
...
+
+### Summary
+
+In this chapter, we learnt about the basics of parameterized tests and how to use them in GoogleTest. We also learnt how to combine test fixture with parameterised tests. Finally, we learnt the advantages of parameterised tests.
```
Comment (dc2917) [inline _episodes/l4_04_parameterized_tests.md]:
```suggestion
In this chapter, we learnt about the basics of parameterised tests and how to use them in GoogleTest. We also learnt how to combine test fixture with parameterised tests. Finally, we learnt the advantages of parameterised tests.
```

File: _episodes/l4_04_parameterized_tests.md (hunk 16 above)
Comment (jamesturner246) [inline _episodes/l4_04_parameterized_tests.md]: To really drive the point home, how about an example?
```suggestion
While the above solution works pretty well, it has a serious drawback. If we carefully look at the tests, we see that the test logic is repeated in both the tests. The only difference between the two tests are the input and output values. Moreover, managing such test will become problematic as the number of test conditions (or input/output) values increases. Imagine if bonus also depended on productivity, experience, age, etcetera? The number of input variations to test grows exponentially as the number and range of arguments grows.
```

File: _episodes/l4_04_parameterized_tests.md [hunk 33]
```diff
@@ -0,0 +1,353 @@
...
+
+```cpp
+// Create a test fixture.
+class EmployeeTestFixture : public::testing::Test {
+    public:
+        Employee employee{"John", 25, 8000, 5, 1000};
+
+};
+
+// Test if we can set the name of an employee.
```
Comment (jamesturner246) [inline _episodes/l4_04_parameterized_tests.md]: This code section is very big. The concern is that it might be difficult and slow to walk though and explain each section of the code by voice to the seminar. Could this be made a bit more concise? Maybe remove some of the repeated boilerplate stuff. A deeper dive into the code here could be a great exercise for the viewer.

**Review by jamesturner246 — CHANGES_REQUESTED** [review]
I think this chapter is especially nice -- well done! the only critique I have is the section with the large block of code at the end. Could be difficult to walk throuhg live and explain, and could maybe be condensed. A deeper dive of the code can (and probably should) be left for a practical exercise,

File: code/Chapter4/2_Test_using_for_loop.cpp (hunk 5 above)
Comment (dalonsoa) [reply code/Chapter4/2_Test_using_for_loop.cpp]: I've been always told that this is bad practice - much like doing in Python `from some_module import *` because you risk clashing names. As it seems that C++ is way more anarchic, let's keep it if you want, but we should be opinionated when trying to encourage what we think are good practices.

File: _episodes/l4_04_parameterized_tests.md [hunk 34]
```diff
@@ -0,0 +1,353 @@
...
+A parameterized tests in GoogleTest requires the following components in general.
+
+1. **A parameterised test class**: Similar to the process of a test fixture, we need to create a class derived from `testing::TestWithParam<T>` where `T` could be any valid C++ type.
+
+```cpp
+class YourTestParameterisedClass : public::TestWithParam<T> {
+    public:
+        ClassUnderTest publicInstance;
+};
+```
```
Comment (dalonsoa) [inline _episodes/l4_04_parameterized_tests.md:136]: So, a question form my side: can you only run parametrised tests in classes and not in top level functions?

File: _episodes/l4_04_parameterized_tests.md (hunk 21 above)
Comment (dalonsoa) [inline _episodes/l4_04_parameterized_tests.md]: This is probably beyond the scope of the course, but cannot you create parametrised tests that use fixtures? Are they mutually exclusive?

Comment (dalonsoa) [reply _episodes/l4_04_parameterized_tests.md]: Ok, you answered this below. I'm too impatient...

File: _episodes/l4_04_parameterized_tests.md (hunk 33 above)
Comment (dalonsoa) [reply _episodes/l4_04_parameterized_tests.md]: I agree with James. It is useful to illustrate a point with small snippets of code, but big chunks of code just make the lesson unreadable. At the very least, remove the first 3 tests (the `TEST_F` ones) as they are not relevant for the current discussion. Also, it might help to break the big chunk of code into one chunk per class/function and take the comments out of the code and into the lesson, in bullet points, something like:

```markdown
1. Create a `EmployeeTestFixture` as done in the previous episode.
2. Create a structure that holds the input and output values. This structure is used to inject values into the test.

```cpp
struct TestValues{
    double inp_salary;
    double inp_bonus;
    double inp_years_employed;
    double out_tax;
    
    TestValues(double salary, double bonus, double years_employed, double tax) 
              : inp_salary(salary), 
                inp_bonus(bonus), 
                inp_years_employed(years_employed), 
                out_tax(tax) {}
};
```

3. Create a parameterised test class from the fixture.

```cpp
class EmployeeTestParameterisedFixture : public EmployeeTestFixture, 
                                         public WithParamInterface<TestValues> {
};
```
```

etc...

**Review by dalonsoa — CHANGES_REQUESTED** [review]
The content and details looks good, but as James and I point out, we feel the amount of code presented in the lesson should be reduced to the bare minimum necessary to illustrate a point - it does not need to be working code. 

Also, it seems that `using namespace :: std;` and `using namespace :: testing;` is not a good practice as it can result in clashes of names. Dan has suggested to remove those from the code and call the relevant functions/classes with the appropriate namespace name.

File: _episodes/l4_04_parameterized_tests.md
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md:1]: As we discussed in PR #14, this is not the case with the current template. So, I am not renaming the file.

File: _episodes/l4_04_parameterized_tests.md (hunk 21 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md]: Yes, we can and as you figured out, it has been explained in the next subsection.

File: code/Chapter4/1_Not_parameterised.cpp (hunk 2 above)
Comment (lkr-ind) [reply code/Chapter4/1_Not_parameterised.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/1_Not_parameterised.cpp (hunk 3 above)
Comment (lkr-ind) [reply code/Chapter4/1_Not_parameterised.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/2_Test_using_for_loop.cpp (hunk 5 above)
Comment (lkr-ind) [reply code/Chapter4/2_Test_using_for_loop.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/2_Test_using_for_loop.cpp (hunk 4 above)
Comment (lkr-ind) [reply code/Chapter4/2_Test_using_for_loop.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/3_Parameterised_not_using_fixture.cpp (hunk 7 above)
Comment (lkr-ind) [reply code/Chapter4/3_Parameterised_not_using_fixture.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/3_Parameterised_not_using_fixture.cpp (hunk 8 above)
Comment (lkr-ind) [reply code/Chapter4/3_Parameterised_not_using_fixture.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/4_param_test_based_fixture.cpp (hunk 9 above)
Comment (lkr-ind) [reply code/Chapter4/4_param_test_based_fixture.cpp]: Resolved in abab5e4457494945e7c2f0fca8e116aed1c8281e

File: code/Chapter4/employee.cpp
Comment (lkr-ind) [reply code/Chapter4/employee.cpp:1]: Yes, that is the same file essentially. Just copied here so that `Chapter 4` folder has all contents at one place.

File: _episodes/l4_04_parameterized_tests.md (hunk 12 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md:13]: Resolved in 62676045679e0e5473c573ba35db49643fcbe7ba

File: _episodes/l4_04_parameterized_tests.md (hunk 16 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md]: Resolved in 62676045679e0e5473c573ba35db49643fcbe7ba

File: _episodes/l4_04_parameterized_tests.md (hunk 33 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md]: @jamesturner246 @dalonsoa 
I agree that the code might be long but I cannot reduce it much. Also, please note
1. We have already discussed most of the concepts in the previous chapter. We do not need to explain each and everything in detail here.
2. I would like to see at least the basic entities that explain how to write parameterised test with fixtures. Most of the code just does that. Extra code has been removed in 02b7b5b7c88133b4f08fd969bb0ff589521b34f9.
3. Even if I separate it into multiple points (as @dalonsoa  mentioned) , it is not going to make code/text shorter. I will not follow this approach. This kind of explanation is useful when we are introducing for the first time as was done in a previous section.  This section is based on the concepts already explained. So, it should be alright to have it one code block (although the overall length has been reduced now).

**Comment by lkr-ind** [comment]
> The content and details looks good, but as James and I point out, we feel the amount of code presented in the lesson should be reduced to the bare minimum necessary to illustrate a point - it does not need to be working code.
> 
> Also, it seems that `using namespace :: std;` and `using namespace :: testing;` is not a good practice as it can result in clashes of names. Dan has suggested to remove those from the code and call the relevant functions/classes with the appropriate namespace name.

This has now been resolved.

File: _episodes/l4_04_parameterized_tests.md (hunk 34 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md:136]: @dalonsoa Thanks for this question. We can run parameterized test in any function and the procedure is similar and straightforward.

Here is the code to illustrate that
```cpp
#include <gtest/gtest.h>
#include <vector>

// Define your normal function
int Sum(int a, int b) {
    return a + b;
}

// Define a test fixture class
class ParameterizedTest : public testing::TestWithParam<std::pair<int, int>> {
};

// Define the test case with the parameterized test
TEST_P(ParameterizedTest, TestSum) {
    // Get the parameter values
    int a = GetParam().first;
    int b = GetParam().second;

    // Call your normal function
    int result = Sum(a, b);

    // Perform assertions
    ASSERT_EQ(a + b, result);
}

// Define the test data
INSTANTIATE_TEST_SUITE_P(Default, ParameterizedTest, testing::Values(
    std::make_pair(1, 1),
    std::make_pair(2, 3),
    std::make_pair(-5, 10)
));

int main(int argc, char** argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
```

File: _episodes/l4_04_parameterized_tests.md [hunk 35]
```diff
@@ -0,0 +1,340 @@
...
+- "How can I write my own parameterised tests using GoogleTest?"
+objectives:
+- "Understand why do we need a parameterised test."
+- "Understand how to create a parameterised test using GoogleTest."
+- "Understand the advantages of parameterised tests."
+- "Understand how to combine a test fixtures with parameterised test."
+keypoints:
+- "We learnt about parameterised tests and how to create them in GoogleTest."
+- "We learnt how parameterised tests are better as compared to using a loop for testing multiple values, conditions, branches etc."
+- "We learnt how to combine test fixture with parameterised test."
```
Comment (dalonsoa) [inline _episodes/l4_04_parameterized_tests.md]: Like in the fixtures lessons, this should be take away messages. I'd suggest the following:

```suggestion
- "Parameterised tests can be used to repeat a specific test with different inputs, reducing code duplication."
- "Parameterised tests are individual tests, so they are more concise and easy to maintain than using a loop for testing multiple conditions"
- "Fixtures can be combined with parameterised tests for maximum flexibility."
```

File: _episodes/l4_04_parameterized_tests.md (hunk 33 above)
Comment (dalonsoa) [reply _episodes/l4_04_parameterized_tests.md]: Ok, point taken. But you still have a `TEST_F` here that is not relevant for this section.

**Review by dalonsoa — APPROVED** [review]
I've left a couple of comments that needs addressing, but it looks really good, otherwise.

File: _episodes/l4_04_parameterized_tests.md (hunk 33 above)
Comment (lkr-ind) [reply _episodes/l4_04_parameterized_tests.md]: Thanks for pointing that out. Resolved in  0520b6bdcc28e7f3813a509e846577b7de62790f

**Comment by lkr-ind** [comment]
@jamesturner246 Can you please let me know if something is remaining. I want to close this PR.

**Review by jamesturner246 — APPROVED** [review]
Thanks Lokesh. Go ahead.
